src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,741 @@
+    bool buffer_failed = false;
+    bool string_failed = false;
+    // io_error has different functionality, it will be set to errno which returned by system
+    // functions (such as read()).
+    int io_error = 0;
+
+    // Internal function to load from file descriptor (a wrapper for buf->fill())
+    int load_into_buf(unsigned len) noexcept;
+
+    // Throw exception based on current state
+    void throw_exception_on(const int states);
+
+    public:
+    istream() noexcept
+    {
+        buf = std::unique_ptr<streambuf>(new(std::nothrow) streambuf);
+        if (!buf) buffer_failed = true;
+    }
+
+    istream(std::unique_ptr<streambuf> passed_buf) noexcept
+    {
+        buf = std::move(passed_buf);
+    }
+
+    istream(const int tfd) noexcept
+    {
+        fd = tfd;
+        buf = std::unique_ptr<streambuf>(new(std::nothrow) streambuf);
+        if (!buf) buffer_failed = true;
+    }
+
+    istream(const int tfd, std::unique_ptr<streambuf> passed_buf) noexcept
+    {
+        fd = tfd;
+        buf = std::move(passed_buf);
+    }
+
+    istream(const istream &) = delete;
+
+    void operator=(const istream &) = delete;
+
+    istream(const istream &&) = delete;
+
+    void operator=(const istream &&) = delete;
+
+    // Open file path which specified in path or as open(2) parameter with flags or modes.
+    //
+    // Note: O_RDONLY is hard-coded in internal call to POSIX open(2).
+    //
+    // Note: Opening an already opened stream is not allowed and caller must close the stream
+    // before open it again.
+    //
+    // Returns: true on success, false on failure.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_system_err on POSIX open(2) failure
+    bool open_nx(const char *path) noexcept;
+    bool open(const char *path);
+
+    bool open_nx(const char *path, const int flags) noexcept;
+    bool open(const char *path, const int flags);
+
+    bool open_nx(const char *path, const int flags, const mode_t mode) noexcept;
+    bool open(const char *path, const int flags, const mode_t mode);
+
+    // Close the current file descriptor.
+    //
+    // Returns: true on success, false on failure cases.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_system_err on system I/O failure
+    bool close_nx() noexcept;
+    bool close();
+
+    // Returns: current bits in stream based on io_states enum.
+    int current_state() noexcept;
+
+    // Returns: true when there is no set bits/vars otherwise false.
+    bool good() noexcept;
+
+    // Returns: true or false based on eofbit is set or not.
+    bool eof() noexcept;
+
+    // Returns: true or false based on buffer_failbit is set or not.
+    bool buffer_failure() noexcept;
+
+    // Returns: true or false based on string_failbit is set or not.
+    bool string_failure() noexcept;
+
+    // Returns: 0 or errno based on io_error value.
+    int io_failure() noexcept;
+
+    // Returns: failure bits/vars to false (or 0 about io_error).
```

Comment:
If the buffer couldn't be allocated, this currently sets `buffer_failed = false` but doesn't actually fix the problem (the buffer still won't be allocated). That seems like an oversight. It would cause a null pointer dereference if you clear the error condition when buffer_failed is true and then try to write to the stream.

Probably you should document that this function doesn't clear the buffer_fail condition (and the behaviour should reflect that).